### ISSUE TITLE
[Response Headers]: set multiple headers of same key in response

### ISSUE THREAD

**Issue by kamleshchandnani** [node-fetch/node-fetch#771]
I've a node-proxy server using node-fetch to get the data from server. The server responds with multiple `set-cookie` header. I want to add this to the response to the client.
I first parsed using `res.headers.raw()['set-cookie']` this gives me array of cookies but now I want to set it in response and I tried using `set`, `append` both of these methods just set the value in one header `set-cookie` separated by comma. Is there any way to have multiple `set-cookie` headers in response?
`set-cookie: cookie1=value1`
`set-cookie: cookie2=value2`

**Comment by dzek69** [comment]
Looks like this is how standard is defined, try this in the browser console and you will get the stringified result too:
```javascript
x = new Response()
x.headers.set("a", "1")
x.headers.append("a", "2")
x.headers.get("a")
```

node-fetch is mostly about following the specs

What's interesting is that documentation on MDN is wrong and suggests an array being returned when appending another header:
https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
```javascript
console.log(myHeaders.get('X-Custom-Header')); // ['ProcessThisImmediately', 'AnotherValue']
```

How to get around this? I have no idea. As you are controlling the response and not what's reading it no "workaround" with splitting string is good here.

Probably using something else, more raw that is not following the Response/Headers standard is the only way to go.

**Comment by sergei-smolnikov** [comment]
I've faced the same issue several times, but had no chance to implement some workaround.
Speaking about specs, the https://tools.ietf.org/html/rfc6265 says:

>    Origin servers SHOULD NOT fold multiple Set-Cookie header fields into
   a single header field.  The usual mechanism for folding HTTP headers
   fields (i.e., as defined in [RFC2616]) might change the semantics of
   the Set-Cookie header field because the %x2C (",") character is used
   by Set-Cookie in a way that conflicts with such folding.

See Section 3 "Overview". There are also examples with multiple `Set-Cookie` headers.

@xxczaki I didn't check the [3.0.0-beta6](https://github.com/node-fetch/node-fetch/blob/master/docs/CHANGELOG.md#v300-beta6) with #834 applied, could you have a look if `Set-Cookie` behavior might changed in v3?

**Comment by smolnikov** [comment]
I've faced the same issue several times, but had no chance to implement some workaround.
Speaking about specs, the https://tools.ietf.org/html/rfc6265 says:

>    Origin servers SHOULD NOT fold multiple Set-Cookie header fields into
   a single header field.  The usual mechanism for folding HTTP headers
   fields (i.e., as defined in [RFC2616]) might change the semantics of
   the Set-Cookie header field because the %x2C (",") character is used
   by Set-Cookie in a way that conflicts with such folding.

See Section 3 "Overview". There are also examples with multiple `Set-Cookie` headers.

@xxczaki I didn't check the [3.0.0-beta6](https://github.com/node-fetch/node-fetch/blob/master/docs/CHANGELOG.md#v300-beta6) with #834 applied, could you have a look if `Set-Cookie` behavior might changed in v3?

**Comment by bitinn** [comment]
AFAIK what you want should be do able, see this test:

https://github.com/node-fetch/node-fetch/blob/master/test/headers.js#L31-L50

I don't think set-cookie headers are handled any differently. But you should try `Cookie` instead of `Set-Cookie`, because nodejs http request module expect that, and we pass headers as such:

https://stackoverflow.com/questions/24365600/passing-cookies-in-nodejs-http-request

(note that Browser's fetch doesn't allow you to set cookies so you won't find a good reference)

If this still fails, send us a PR with a failing test and we can look into it.

**Comment by smolnikov** [comment]
@bitinn my use case is related to `Response`, not `Request`. I actually came to `node-fetch` from `apollo-server` (that [is using](https://github.com/apollographql/apollo-server/blob/master/packages/apollo-server-env/src/polyfills/fetch.js) `node-fetch`) where I'm sending the cookies to the client with a custom plugin. So I need to add several `Set-Cookie` headers `Response`.
In my case, the cookies should be all `HttpOnly`, so there's no other way to send them to the client.

**Comment by jamesarosen** [comment]
Here's a clear demonstration of the bug.

HTTPBin sets cookie according to the spec (one cookie per `set-cookie` header):

```sh
$ curl -svo /dev/null 'https://www.github.com' 2>&1 | grep -i Set-Cookie
Set-Cookie: _gh_sess=b8z...; Path=/; HttpOnly; Secure; SameSite=Lax
Set-Cookie: _octo=GH1.1.1...; Path=/; Domain=github.com; Expires=Sun, 18 Jul 2021 02:42:55 GMT; Secure; SameSite=Lax
Set-Cookie: logged_in=no; Path=/; Domain=github.com; Expires=Sun, 18 Jul 2021 02:42:55 GMT; HttpOnly; Secure; SameSite=Lax
```

```js
const fetch = require('node-fetch')
fetch("https://www.github.com").then(res => { console.log(res.headers.get('set-cookie')) })
_gh_sess=b8z...; Path=/; HttpOnly; Secure; SameSite=Lax, _octo=GH1.1.1...; Path=/; Domain=github.com; Expires=Sun, 18 Jul 2021 02:42:55 GMT; Secure; SameSite=Lax, logged_in=no; Path=/; Domain=github.com; Expires=Sun, 18 Jul 2021 02:42:55 GMT; HttpOnly; Secure; 
```

There's no good way to parse those back out. You can't split on `,` because of directives like `Expires=Sun, 18 Jul 2021 02:42:55 GMT`.

**Comment by benmccann** [comment]
@tinovyatkin @xxczaki while I'm not 100% sure, it looks to me like this is a real bug and deviation from the spec. I was wondering if you had thoughts on this before the v3 API is finalized? Should we switch the type of headers from `string` to `string | string[]` to match Node's API?

**Comment by benmccann** [comment]
It looks like `Set-Cookie` is not accessible from JS according to the `fetch` spec: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie

**Comment by dzek69** [comment]
Yes @benmccann, but node-fetch aims to replicate fetch API inside node, not to being a polyfill or something for the browsers. Setting a cookie when sending a request from a server is very useful. See readme for more details.

So this limitation isn't a problem when trying to solve this issue.

**Comment by jimmywarting** [comment]
I happen to stumble up on https://github.com/nfriedly/set-cookie-parser
it has a [method](https://github.com/nfriedly/set-cookie-parser/blob/62b29d5498bf4337c8d10ed29aa971326efb22eb/lib/set-cookie.js#L126-L197) to split comma separated values
```js
var cookies = setCookie.splitCookiesString(headers.get('set-cookie'))

// or
var cookie = '_gh_sess=b8z...; Path=/; HttpOnly; Secure; SameSite=Lax, _octo=GH1.1.1...; Path=/; Domain=github.com; Expires=Sun, 18 Jul 2021 02:42:55 GMT; Secure; SameSite=Lax, logged_in=no; Path=/; Domain=github.com; Expires=Sun, 18 Jul 2021 02:42:55 GMT; HttpOnly; Secure;'
var cookies = setCookie.splitCookiesString(cookie))
console.log(cookies) 
[
  "_gh_sess=b8z...; Path=/; HttpOnly; Secure; SameSite=Lax",
  "_octo=GH1.1.1...; Path=/; Domain=github.com; Expires=Sun, 18 Jul 2021 02:42:55 GMT; Secure; SameSite=Lax",
  "logged_in=no; Path=/; Domain=github.com; Expires=Sun, 18 Jul 2021 02:42:55 GMT; HttpOnly; Secure; "
]
```

**Comment by dzek69** [comment]
@jimmywarting typ, but this is still incorrect to do: https://github.com/node-fetch/node-fetch/issues/771#issuecomment-635968570

but if native fetch Reponse is working like that: https://github.com/node-fetch/node-fetch/issues/771#issuecomment-613281859

I guess that's the only option now...

**Comment by benjaminjkraft** [comment]
I want to highlight what @smolnikov said above -- RFC 6265 says one SHOULD NOT do this *for `Set-Cookie` specifically* (and RFC 7230 echoes this as a special case).  It's acceptable for other headers, just not `Set-Cookie`.  The browser `fetch` doesn't need to care because it doesn't let you see `Set-Cookie` at all.

That said, for anyone else reading this thread, it's easy enough to work around by using `headers.raw()['Set-Cookie']`, which will give you a proper array of set-cookie values.  That's what we're doing.

**Comment by yhojann-cl** [comment]
> That said, for anyone else reading this thread, it's easy enough to work around by using `headers.raw()['Set-Cookie']`

1. The headers is not case sensitive but for accesing to an array is case sensitive.
2. The `raw()` function can not return a native array, you can not call `raw().filter(h => ...)`. Return a object as semi array :-/
3. When cast to native array get an empty array, by example: `Array.prototype.slice.call(response.headers.raw())`

You need:

```javascript
if(Object.keys(response.headers.raw()).includes('set-cookie')) {
    console.log(response.headers.raw()['set-cookie']);
}
```

**Comment by zargold** [comment]
It seems that headers.raw() is no longer valid in Node 18? Not sure but I got a run-time error when running in Node-18?

Previously working (in Node-16):
``` javascript
const cookiesSet = response.headers.raw()['set-cookie'];
```
Now:
```
response.headers.raw is not a function","stack":"TypeError: response.headers.raw is not a function\n at this line
```
